src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/AssemblyBuilderImpl.cs

```diff
@@ -0,0 +1,129 @@
+        {
+            ArgumentNullException.ThrowIfNull(name);
+
+            _assemblyName = name;
+
+            if (assemblyAttributes != null)
+            {
+                foreach (CustomAttributeBuilder assemblyAttribute in assemblyAttributes)
+                {
+                    SetCustomAttribute(assemblyAttribute);
+                }
+            }
+        }
+
+        internal static AssemblyBuilderImpl DefineDynamicAssembly(AssemblyName name)
+            => new AssemblyBuilderImpl(name, null);
+
+        internal static AssemblyBuilderImpl DefineDynamicAssembly(
+            AssemblyName name,
+            IEnumerable<CustomAttributeBuilder>? assemblyAttributes)
+                => new AssemblyBuilderImpl(name, assemblyAttributes);
+
+        private static void WritePEImage(Stream peStream, MetadataBuilder metadataBuilder, BlobBuilder ilBuilder) // MethodDefinitionHandle entryPointHandle when we have main method.
+        {
+            // Create executable with the managed metadata from the specified MetadataBuilder.
+            var peHeaderBuilder = new PEHeaderBuilder(
+                imageCharacteristics: Characteristics.Dll // Start off with a simple DLL
+                );
+
+            var peBuilder = new ManagedPEBuilder(
+                peHeaderBuilder,
+                new MetadataRootBuilder(metadataBuilder),
+                ilBuilder);
+
+            // Write executable into the specified stream.
+            var peBlob = new BlobBuilder();
+            peBuilder.Serialize(peBlob);
+            peBlob.WriteContentTo(peStream);
+        }
+
+        internal void Save(Stream stream)
+        {
+            ArgumentNullException.ThrowIfNull(stream);
+
+            if (_module == null)
+            {
+                throw new InvalidOperationException(SR.AModuleRequired);
+            }
+
+            if (_previouslySaved) // Cannot save an assembly multiple times. This is consistent with Save() in .Net Framework.
+            {
+                throw new InvalidOperationException(SR.CannotSaveMultipleTimes);
+            }
+
+            if (string.IsNullOrEmpty(_assemblyName.Name))
+            {
+                throw new InvalidOperationException(SR.AssemblyNameCannotBeNull);
+            }
+
+            // Add assembly metadata
+            var metadata = new MetadataBuilder();
+            metadata.AddAssembly(
+               metadata.GetOrAddString(value: _assemblyName.Name),
+               version: _assemblyName.Version ?? new Version(0, 0, 0, 0),
+               culture: _assemblyName.CultureName == null ? default : metadata.GetOrAddString(value: _assemblyName.CultureName),
+               publicKey: _assemblyName.GetPublicKey() is byte[] publicKey ? metadata.GetOrAddBlob(value: publicKey) : default,
+               flags: (AssemblyFlags)_assemblyName.Flags,
+               hashAlgorithm: AssemblyHashAlgorithm.None); // AssemblyName.HashAlgorithm is obsolete so default value used.
```

Comment:
```suggestion
               hashAlgorithm: _assemblyName.HashAlgorithm);
```
It is obsolete, but that's not a good enough reason to ignore it.